File: examples/hosted-checkout-typescript/src/order-store.ts
export type OrderStatus = "created" | "pending" | "paid" | "fulfilled_unsettled" | "review_required";

export interface Order {
  id: string;
  amount_minor: number;
  currency: "JPY" | "USD";
  payment_attempt: number;
  siglume_challenge_hash?: string;
  siglume_checkout_session_id?: string;
  siglume_requirement_id?: string;
  siglume_chain_receipt_id?: string;
  siglume_payment_status: OrderStatus;
}

const orders = new Map<string, Order>([
  [
    "order_123",
    {
      id: "order_123",
      amount_minor: 1200,
      currency: "JPY",
      payment_attempt: 0,
      siglume_payment_status: "created",
    },
  ],
]);

const processedWebhookEvents = new Set<string>();

export function getOrder(orderId: string): Order | undefined {
  return orders.get(orderId);
}

export function allOrders(): Order[] {
  return [...orders.values()];
}

export function saveOrder(order: Order): void {
  orders.set(order.id, order);
}

export function findOrderByChallengeHash(challengeHash: string): Order | undefined {
  return [...orders.values()].find((order) => order.siglume_challenge_hash === challengeHash);
}

export function markWebhookEventProcessedOnce(eventId: string): boolean {
  if (processedWebhookEvents.has(eventId)) {
    return false;
  }
  processedWebhookEvents.add(eventId);
  return true;
}

File: examples/hosted-checkout-python/order_store.py
from __future__ import annotations

from typing import Any

Order = dict[str, Any]

_orders: dict[str, Order] = {
    "order_123": {
        "id": "order_123",
        "amount_minor": 1200,
        "currency": "JPY",
        "payment_attempt": 0,
        "siglume_payment_status": "created",
    }
}
_processed_webhook_events: set[str] = set()


def get_order(order_id: str) -> Order | None:
    return _orders.get(order_id)


def begin_checkout_attempt(order_id: str) -> Order | None:
    order = _orders.get(order_id)
    if order is None:
        return None
    if not int(order.get("payment_attempt") or 0):
        order["payment_attempt"] = 1
    return order


def all_orders() -> list[Order]:
    return list(_orders.values())


def save_order(order: Order) -> None:
    _orders[str(order["id"])] = order


def find_order_by_challenge_hash(challenge_hash: str) -> Order | None:
    for order in _orders.values():
        if order.get("siglume_challenge_hash") == challenge_hash:
            return order
    return None


def process_webhook_event_once(event_id: str, handler) -> str:
    if event_id in _processed_webhook_events:
        return "duplicate"
    handler()
    _processed_webhook_events.add(event_id)
    return "processed"
